src/main/java/com/amazon/opendistroforelasticsearch/sql/executor/format/DateFieldFormatter.java

```diff
@@ -0,0 +1,167 @@
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.commons.lang3.time.DateUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.text.ParseException;
+import java.util.Collection;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Formatter to transform date fields into a consistent format for consumption by clients.
+ */
+public class DateFieldFormatter {
+    private static final Logger LOG = LogManager.getLogger(DateFieldFormatter.class);
+    private static final String FORMAT_JDBC = "yyyy-MM-dd HH:mm:ss.SSS";
+
+    private static final String FORMAT_DOT_DATE_AND_TIME = "yyyy-MM-dd'T'HH:mm:ss.SSSZ";
+    private static final String FORMAT_DOT_KIBANA_SAMPLE_DATA_LOGS_EXCEPTION = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'";
+    private static final String FORMAT_DOT_KIBANA_SAMPLE_DATA_FLIGHTS_EXCEPTION = "yyyy-MM-dd'T'HH:mm:ss";
+    private static final String FORMAT_DOT_KIBANA_SAMPLE_DATA_ECOMMERCE_EXCEPTION = "yyyy-MM-dd'T'HH:mm:ssXXX";
+    private static final String FORMAT_DOT_DATE = "yyyy-MM-dd";
+
+    private final Map<String, String> dateFieldFormatMap;
+    private final Map<String, String> fieldAliasMap;
+    private Set<String> dateColumns;
+
+    public DateFieldFormatter(String indexName, List<Schema.Column> columns, Map<String, String> fieldAliasMap) {
+        this.dateFieldFormatMap = getDateFieldFormatMap(indexName);
+        this.dateColumns = getDateColumns(columns);
+        this.fieldAliasMap = fieldAliasMap;
+    }
+
+    @VisibleForTesting
+    protected DateFieldFormatter(Map<String, String> dateFieldFormatMap,
+                                 List<Schema.Column> columns,
+                                 Map<String, String> fieldAliasMap) {
+        this.dateFieldFormatMap = dateFieldFormatMap;
+        this.dateColumns = getDateColumns(columns);
+        this.fieldAliasMap = fieldAliasMap;
+    }
+
+    /**
+     * Apply the JDBC date format ({@code yyyy-MM-dd HH:mm:ss.SSS}) to date values in the current row.
+     *
+     * @param rowSource The row in which to format the date values.
+     */
+    public void applyJDBCDateFormat(Map<String, Object> rowSource) {
+        for (String columnName : dateColumns) {
+            Object columnOriginalDate = rowSource.get(columnName);
+            if (columnOriginalDate == null) {
+                // Don't try to parse null date values
+                continue;
+            }
+
+            String columnFormat = getFormatForColumn(columnName);
+            if (columnFormat == null) {
+                LOG.warn("Could not determine date format for column {}; returning original value", columnName);
+                continue;
+            }
+            DateFormat format = DateFormat.valueOf(columnFormat.toUpperCase());
```

Comment:
Is it possible the columnFormat is not included in the DateFormat? For example, custom-date-formats, https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-date-format.html#custom-date-formats

Reply:
Yeah, this does look to be a possible case. Will add handling & testing for this

Reply:
Updated implementation & added tests for handling custom formats